src/main/controller/PromotionController.ts

```diff
@@ -183,12 +185,50 @@ export class PromotionController {
   ): Promise<any> => {
     try {
       await getManager().transaction(async (transactionalEntityManager) => {
-        const id = await IdValidation.schema.validateAsync(request.params.id, {
+        const pid = await IdValidation.schema.validateAsync(request.params.id, {
+          abortEarly: false,
+        });
+        const uid = await IdValidation.schema.validateAsync(request.body.uid, {
           abortEarly: false,
         });
+
+        await this.checkIfUserAndPromotionExist(
+          transactionalEntityManager,
+          pid,
+          uid
+        );
+
+        let voteRecord;
+        try {
+          voteRecord = await transactionalEntityManager
+            .getCustomRepository(VoteRecordRepository)
+            .findOneOrFail({ userId: uid, promotionId: pid });
+        } catch (e) {
+          // user vote first time
+          voteRecord = await transactionalEntityManager
+            .getCustomRepository(VoteRecordRepository)
+            .create({ userId: uid, promotionId: pid });
+          await transactionalEntityManager
+            .getCustomRepository(VoteRecordRepository)
+            .save(voteRecord);
+        }
+
+        if (voteRecord.voteState === VoteState.UP) {
+          // user cannot vote twice
+          return response.status(204).send();
+        } else {
+          // user vote first time
+          await transactionalEntityManager
+            .getCustomRepository(PromotionRepository)
+            .increment({ id: pid }, 'votes', 1);
+        }
+        voteRecord.voteState =
+          voteRecord.voteState === VoteState.INIT
+            ? VoteState.UP
+            : VoteState.INIT;
```

Comment:
I was thinking something along the lines of this for how it would flow:
![image](https://user-images.githubusercontent.com/49849754/113256759-803dcb00-927e-11eb-90ab-6f2934e285b0.png)

- For example in case 2, if the user has already downvoted the promotion, they can either:
    - upvote it which will highlight the up arrow 
    - click downvote again to remove their vote (which is setting the vote state to INIT)

Because of this, I was thinking we would need three endpoints
`/upvote` 
`/downvote`
`/resetVote` 

So I was thinking `/upvote` endpoint would follow this logic:
```
1. Find if a vote record exists
    1.  If the vote state is DOWN, set it to UP and increase votes by 2
    2.  If already UP, don't do anything and return
    3. if INIT set it to UP and increase votes by 1
2. If vote record does not exist
    1. Call repository.save({ userId: uid, promotionId: pid, voteState: VoteState.UP})
    2. Increase the number of votes by 1
```

This is an example of reddit where you can see they increase votes by 2 when I jump from down to up
![RedditVoteSystem](https://user-images.githubusercontent.com/49849754/113258157-6bfacd80-9280-11eb-896c-5b24ae831d6d.gif)

Reply:
Thank you for your instruction, it is so much helpful to implement the voting flow. I'm just curious why we need `resetVote`??

Reply:
I see, yah actually we could make the upvote endpoint set the state to INIT if the vote record exists and the state was UP.

I was thinking of having it as a separate endpoint like 'resetVote' because if we make a request to 'upvote', I don't really expect the vote to go to INIT. Then it isn't really clear that the upvote endpoint is actually upvoting.

It's totally up to you though, this is a design decision. Actually, one thing I'd recommend is to see how reddit is doing it by inspecting the networks tab and seeing what they send in the request body or if it's 2/3 different endpoints or all one endpoint

Reply:
Ok, so I gonna follow the logic you gave (how reddit is doing) and then let you know after that!

Reply:
I saw how reddit is doing and it seems to use the same endpoint for up and down vote. It provides the form data with the request which includes `id` and `dir` where `dir` might have `1` , `0` or `-1`. If the status becomes `UP` (INIT => UP or DOWN => UP), the dir is `1`. If the status becomes `INIT` (UP => INIT or DOWN => INIT), the dir is `0`. If the status becomes `DOWN`, the dir is `-1`.

Reply:
Very neat! Hmm I see, so I'm guessing that for this case:

> If the status becomes UP (INIT => UP or DOWN => UP), the dir is 1

If it's INIT => UP, then we add 1 to total votes
If it's DOWN => UP, then we add 2 to total votes
So at the controller level, we would need to know the previous VoteState (if it exists)

I see, either way, I think we can be very flexible with how we approach it and I'll let you choose which one you think is the best!